umbraco/Umbraco-CMS
# Conditionally enabling the delivery API + add protection and preview support

### Prerequisites


### Description

This PR adds the ability to explicitly enable the delivery API and also adds public access protection and preview support. Both of these features are managed through a new `ContentApi` configuration section in `appsettings.json`.

As part of the implementation I have refactored all services to be singletons: I have also ensured no-op implementations for all required services, so the delivery API does not throw exceptions when it is not added explicitly to the composition of the running site.

### Testing this API

1. Remove `.AddContentApi()` from `ConfigureServices`.
![image](https://user-images.githubusercontent.com/7405322/226573610-b4c894a9-0cb8-4e9c-9761-ead9f69c3c5c.png)
2. Request one of the API endpoints. Verify that you'll get a nice 401 response and not a server error.
![image](https://user-images.githubusercontent.com/7405322/226574020-f91dbbca-d361-4d7c-b359-3ca85ef70c5a.png)
3. Re-add `.AddContentApi()` to `ConfigureServices`.
4. Request the API endpoint again. Verify that you'll *still* get a 401 response.
5. Add the following configuration to `appsettings.json` under `Umbraco:CMS`:
```json
"ContentApi": {
  "Enabled": true,
  "PublicAccess": true,
  "ApiKey": "my-api-key"
}
```
6. Request the API endpoint again. Verify that you'll now get the expected 200 response.
7. Turn public access off by configuration (`"PublicAccess": false`).
8. Request the API endpoint again. Verify that you'll once more get a 401 response.
9. Add an `Api-Key` request header containing the configured API key.
10. Request the API endpoint again. Verify that you'll now get the expected 200 response.
11. Turn public access back on by configuraion (`"PublicAccess": true`).
12. Remove the API key request header.
13. Add a `Preview` request header with the value `true`.
14. Request the API endpoint again. Verify that you'll once more get a 401 response.
15. Re-add the `Api-Key` request header.
16. Request the API endpoint again. Verify that you'll now see draft content in the response.
